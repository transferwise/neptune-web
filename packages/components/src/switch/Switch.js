import React from 'react';
import Types from 'prop-types';
import classnames from 'classnames';
import { CheckCircle, CrossCircle } from '@transferwise/icons';
import './Switch.css';
import { logActionRequiredIf } from '../utilities';

import KeyCodes from '../common/keyCodes';

const Switch = (props) => {
  const { checked, className, id, onClick } = props;
  const handleKeyDown = (event) => {
    if (event.code === 32 || event.keyCode === KeyCodes.SPACE) {
      event.preventDefault();
      onClick();
    }
  };

  const ariaLabel = props['aria-label'];
  const ariaLabelledby = ariaLabel ? null : props['aria-labelledby'];

  logActionRequiredIf(
    'Switch now expects either `aria-label` or `aria-labelledby`, and will soon make these props required. Please update your usage to provide one or the other.',
    !ariaLabel && !ariaLabelledby,
  );

  return (
    <span
      className={classnames(
        'np-switch',
        {
          'np-switch--unchecked': !checked,
          'np-switch--checked': checked,
        },
        className,
      )}
      onClick={onClick}
      onKeyDown={handleKeyDown}
      tabIndex={0}
      role="switch"
      aria-checked={checked}
      aria-label={ariaLabel}
      aria-labelledby={ariaLabelledby}
      id={id}
    >
      <span className="np-switch--thumb">
        {checked ? <CheckCircle filled size={24} /> : <CrossCircle filled size={24} />}
      </span>
      <input type="checkbox" checked={checked} readOnly />
    </span>
  );
};

Switch.propTypes = {
  /** Used to describe the purpose of the switch. To be used if there is no external label (i.e. aria-labelledby is null) */
  'aria-label': Types.string,
  /** A reference to a label that describes the purpose of the switch. Ignored if aria-label is provided */
  'aria-labelledby': Types.string,
  /** Whether the switch is checked or not */
  checked: Types.bool,
  /** Classes to apply to the switch container */
  className: Types.string,
  /** ID to apply to the switch container */
  id: Types.string,
  /** Function called when the switch is toggled */
  onClick: Types.func.isRequired,
};

Switch.defaultProps = {
  'aria-label': null,
  'aria-labelledby': null,
  checked: false,
  className: null,
  id: null,
};

export default Switch;
