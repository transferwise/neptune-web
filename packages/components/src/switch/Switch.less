@import (reference) '@transferwise/neptune-css/src/variables/neptune-tokens.less';

.np-switch {
  display: inline-flex;
  overflow: hidden;
  width: 50px;
  padding: 2px;
  cursor: pointer;
  border: 1px solid @color-base-smoke-mid;
  border-radius: 16px;

  &--thumb {
    display: flex;
    transition: transform cubic-bezier(0, 0.94, 0.62, 1) 350ms;

    .tw-icon {
      color: var(--color-text-control);
    }
  }

  &--unchecked {
    background: @color-base-smoke-mid;

    .switch--thumb {
      transform: translateX(0);
    }
  }

  &--checked {
    border: 1px solid var(--color-control-accent);
    background: var(--color-control-accent);

    .np-switch--thumb {
      transform: translateX(20px);
    }
  }

  input {
    position: absolute;
    left: -100%;
    display: none;
    width: 0;
    height: 0;
    opacity: 0;
  }
}
