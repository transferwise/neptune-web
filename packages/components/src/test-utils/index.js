import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from '..';
import { DEFAULT_LOCALE } from '../common/locale';
import en from '../../i18n/en.json';

/**
 * Custom `render` function which wraps passed elements in Provider component
 * Source: https://testing-library.com/docs/example-react-intl/#creating-a-custom-render-function
 */
function customRender(ui, { locale = DEFAULT_LOCALE, messages = en, ...renderOptions } = {}) {
  // eslint-disable-next-line react/prop-types
  function Wrapper({ children }) {
    return <Provider i18n={{ locale, messages }}>{children}</Provider>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from '@testing-library/react';
export { userEvent };
export { customRender as render };
