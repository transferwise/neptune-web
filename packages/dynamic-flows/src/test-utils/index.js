import React from 'react';
import { mount } from 'enzyme';
import { Provider } from '@transferwise/components';
import { act } from 'react-dom/test-utils';

const getMockFetchPromise = (status, jsonFn, delay, signal) => {
  const response = {
    status,
    json: jsonFn,
  };
  return new Promise((resolve) => {
    let aborted = false;
    if (signal) {
      signal.addEventListener('abort', () => {
        aborted = true;
      });
    }
    setTimeout(() => {
      return !aborted && resolve(response);
    }, delay);
  });
};

const wait = (t) => {
  return act(() => {
    return new Promise((resolve) => setTimeout(resolve, t));
  });
};

function mountWithProviders(component, locale = 'en-GB', messages = {}) {
  return mount(<Provider i18n={{ locale, messages }}>{component}</Provider>);
}

export { mountWithProviders as mount, getMockFetchPromise, wait };
