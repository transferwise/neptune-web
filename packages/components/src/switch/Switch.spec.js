import React from 'react';
import { render, fireEvent, screen } from '../test-utils';
import Switch from './Switch';

import KeyCodes from '../common/keyCodes';

const props = {
  'aria-label': 'a label',
  checked: true,
  className: 'a-class-name',
  id: 'id',
  onClick: jest.fn(),
};

describe('Switch', () => {
  it('renders component when checked', () => {
    const { container } = render(
      <Switch
        aria-label={props['aria-label']}
        checked={props.checked}
        className={props.className}
        id={props.id}
        onClick={props.onClick}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders component associated with label', () => {
    render(
      <>
        <label id="labelID" htmlFor={props.id}>
          A switch with a label
        </label>
        <Switch
          aria-labelledby="labelID"
          checked={props.checked}
          className={props.className}
          id={props.id}
          onClick={props.onClick}
        />
        ,
      </>,
    );
    expect(screen.getByLabelText('A switch with a label')).toBeInTheDocument();
  });

  it('renders component when unchecked', () => {
    const { container } = render(
      <Switch
        checked={false}
        onClick={props.onClick}
        className={props.className}
        id={props.id}
        aria-label={props['aria-label']}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('calls onClick when user press space key', () => {
    const { getAllByRole } = render(
      <Switch
        checked={props.checked}
        onClick={props.onClick}
        className={props.className}
        id={props.id}
        aria-label={props['aria-label']}
      />,
    );

    const input = getAllByRole('checkbox')[0];
    fireEvent.keyDown(input, { key: '33', keyCode: KeyCodes.ENTER });
    expect(props.onClick).not.toHaveBeenCalled();
    fireEvent.keyDown(input, { key: '32', keyCode: KeyCodes.SPACE });
    expect(props.onClick).toHaveBeenCalled();
  });
});
