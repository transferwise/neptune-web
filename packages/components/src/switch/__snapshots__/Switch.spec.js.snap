// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders component when checked 1`] = `
<div>
  <span
    aria-checked="true"
    aria-label="a label"
    class="np-switch np-switch--checked a-class-name"
    id="id"
    role="switch"
    tabindex="0"
  >
    <span
      class="np-switch--thumb"
    >
      <span
        aria-hidden="true"
        class="tw-icon tw-icon-check-circle "
        role="presentation"
      >
        <svg
          fill="currentColor"
          focusable="false"
          height="24"
          width="24"
        >
          <path
            clip-rule="evenodd"
            d="M12 22c5.523 0 10-4.477 10-10S17.523 2 12 2 2 6.477 2 12s4.477 10 10 10zM7.707 11.293l-1.414 1.414 4.207 4.207 7.707-7.707-1.414-1.414-6.293 6.293-2.793-2.793z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </span>
    <input
      checked=""
      readonly=""
      type="checkbox"
    />
  </span>
</div>
`;

exports[`Switch renders component when unchecked 1`] = `
<div>
  <span
    aria-checked="false"
    aria-label="a label"
    class="np-switch np-switch--unchecked a-class-name"
    id="id"
    role="switch"
    tabindex="0"
  >
    <span
      class="np-switch--thumb"
    >
      <span
        aria-hidden="true"
        class="tw-icon tw-icon-cross-circle "
        role="presentation"
      >
        <svg
          fill="currentColor"
          focusable="false"
          height="24"
          width="24"
        >
          <path
            clip-rule="evenodd"
            d="M12 22c5.523 0 10-4.477 10-10S17.523 2 12 2 2 6.477 2 12s4.477 10 10 10zm3.293-5.293L12 13.414l-3.293 3.293-1.414-1.414L10.586 12 7.293 8.707l1.414-1.414L12 10.586l3.293-3.293 1.414 1.414L13.414 12l3.293 3.293-1.414 1.414z"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </span>
    <input
      readonly=""
      type="checkbox"
    />
  </span>
</div>
`;
